### PR TITLE
Ghost animation for quote list and details overview

### PR DESCRIPTION
**Pull request by Larisa-Staroverova**

### CODE REVIEW

File: feature-libs/quote/components/details/cart/quote-details-cart.component.ts
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/cart/quote-details-cart.component.ts:1]: was the change of the order of the properties intended?

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts [hunk 1]
```diff
@@ -39,10 +39,10 @@ export class QuoteDetailsCommentComponent {
   expandComments = true;
   iconTypes = ICON_TYPE;
 
-  quoteDetails$ = this.quoteFacade.getQuoteDetails();
+  quoteDetails$: Observable<Quote> = this.quoteFacade.getQuoteDetails();
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/comment/quote-details-comment.component.ts:42]: this type declaration is superfluous, as it is implicitly added from the assignmnet

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts [hunk 2]
```diff
@@ -97,7 +97,7 @@ export class QuoteDetailsCommentComponent {
   showComments(quote: Quote): boolean {
     let numItemComments = 0;
     quote.entries?.forEach(
-      (entry) => (numItemComments += entry.comments?.length ?? 0)
+      (entry: OrderEntry) => (numItemComments += entry.comments?.length ?? 0)
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/comment/quote-details-comment.component.ts]: same here - no need to add OrderEntry type - its implicit

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts [hunk 3]
```diff
@@ -112,7 +112,7 @@ export class QuoteDetailsCommentComponent {
         this.quoteUiConfig.quote?.maxCharsForComments ??
         DEFAULT_COMMENT_MAX_CHARS,
       displayAddMessageSection: this.quoteDetails$.pipe(
-        map((quote) => quote.isEditable)
+        map((quote) => quote?.isEditable)
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/comment/quote-details-comment.component.ts]: why did you add the optional chaining operator here?  - its guaranteed here that quote is neither undefined nor null. please revert this. And also for all other occurrences in the file.

File: feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts [hunk 4]
```diff
@@ -127,6 +127,31 @@ describe('QuoteDetailsCommentComponent', () => {
     expect(component).toBeTruthy();
   });
 
+  describe('Ghost animation', () => {
+    it('should render view for ghost animation', () => {
+      component.quoteDetails$ = of(null);
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts]: strict mode issue - null is not valid here

File: feature-libs/quote/components/details/cart/quote-details-cart.component.spec.ts [hunk 5]
```diff
@@ -82,6 +103,94 @@ describe('QuoteDetailsCartComponent', () => {
     expect(component).toBeTruthy();
   });
 
+  describe('Ghost animation', () => {
+    it('should render view for ghost animation', () => {
+      component.quoteDetails$ = of(null);
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/cart/quote-details-cart.component.spec.ts]: strict mode issue - null is not valid here

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts [hunk 6]
```diff
@@ -525,6 +559,7 @@ describe('QuoteActionsByRoleComponent', () => {
         })
       ).toEqual('btn-secondary');
     });
+
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts]: Is there a coding guideline, that enforces us to enter an empty line after each test? - it makes sense if the tests are longer and also have empty lines, but for these mini tests with only 1-3 lines of code?

File: feature-libs/quote/components/quote-list/quote-list.component.spec.ts [hunk 7]
```diff
@@ -34,22 +34,26 @@ const mockPagination: PaginationModel = {
...
   pagination: mockPagination,
   quotes: [mockQuote],
 };
+
 const mockQuoteListState: QueryState<QuoteList> = {
   loading: false,
   error: false,
   data: mockQuoteList,
 };
+
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/quote-list/quote-list.component.spec.ts:56]: is there a coding guideline that enforces us to enter an empty line after each const declaration?

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.ts [hunk 8]
```diff
@@ -55,7 +53,7 @@ export class QuoteActionsByRoleComponent implements OnInit, OnDestroy {
   ngOnInit(): void {
     //submit button present and threshold not reached: Display message
     this.quoteDetails$.pipe(take(1)).subscribe((quote) => {
-      const mustDisableAction = quote.allowedActions.find((action) =>
+      const mustDisableAction = quote?.allowedActions?.find((action) =>
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.ts]: why did you add optional chaining operators here? - quote is guaranteed not to be null or undefined here. also allowedActions are non optional in type quote. please revert.

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]
please check regarding strict mode issues in the tests. (also those that i did not commented on)
please remove all optional chaining operators (.?), that are not required. (also those that i did not commented on)
please also try to limit a bit the formatting changes and refactorings to code that relates to the BI (ghost animations). So we can focus on this. It also makes the review easier, as one sees the relevant changes.

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts (hunk 1 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.ts:42]: I see what you mean but what about `messageEvents$` and `messagingConfigs` that were added by you.
image
I think we have to be consistent that's why I added the missing declaration. 
I have to quote @ChristophHi , if you see points, you should correct them.
By the way, as you could see in the `product-configurator` lib, we always added the type: 
image

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts (hunk 2 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.ts]: Done!

Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.ts]: Because of this: 
image

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts (hunk 3 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.ts]: Done

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts (hunk 6 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts]: I added an empty line between the test for readability purposes.
It's tiring and hard to find and understand something over time when the tests come in a row without a blank line

File: feature-libs/quote/components/quote-list/quote-list.component.spec.ts (hunk 7 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-list/quote-list.component.spec.ts:56]: It's tiring and hard to find and understand something over time when the tests come in a row without a blank line.

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.ts (hunk 8 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.ts]: I was misled with the 12th line:
image
I missed the 37th line.

Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.ts]: Done!

File: feature-libs/quote/components/quote-list/quote-list.component.spec.ts (hunk 7 above)
Comment (Uli-Tiger) [reply feature-libs/quote/components/quote-list/quote-list.component.spec.ts:56]: its not about tests, but about multiline const delcarations....
It's tiring and hard to find and understand something over time when everything is separated with a blank line.

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts (hunk 1 above)
Comment (Uli-Tiger) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.ts:42]: > I see what you mean but what about `messageEvents$` and `messagingConfigs` that were added by you.

I agree the types could be removed there as well to be consistent. in your screenshot, you can se that the variable router data in the pipe also is declared without explicit type.
However we are now discussing code style, which is completely unrelated to ghost animations....
We can't reformat the whole code base in every PR. Hence I suggest to focus on the ghost animations.

File: feature-libs/quote/components/quote-list/quote-list.component.spec.ts [hunk 9]
```diff
@@ -34,22 +34,26 @@ const mockPagination: PaginationModel = {
   pageSize: 5,
   sort: 'byCode',
 };
+
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/quote-list/quote-list.component.spec.ts:37]: please remove

File: feature-libs/quote/components/quote-list/quote-list.component.spec.ts [hunk 10]
```diff
@@ -34,22 +34,26 @@ const mockPagination: PaginationModel = {
...
 const mockAction = { type: QuoteActionType.EDIT, isPrimary: true };
 const mockQuote: Quote = {
   ...createEmptyQuote(),
   allowedActions: [mockAction],
   cartId: mockCartId,
   code: '333333',
 };
+
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/quote-list/quote-list.component.spec.ts:45]: please remove

File: feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts [hunk 11]
```diff
@@ -166,6 +191,7 @@ describe('QuoteDetailsCommentComponent', () => {
     quote.comments = [];
     quote.comments.push({});
     quote.comments.push({});
+
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts]: agreed: please revert the addition of this empty line, as well as all of the following empty line additions.

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts (hunk 1 above)
Comment (Uli-Tiger) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.ts:42]: agreed: leave you change as it is.

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts [hunk 12]
```diff
@@ -130,7 +130,7 @@ export class QuoteDetailsCommentComponent {
     return this.quoteDetails$.pipe(
       map((quote) => {
         const itemList: Item[] = [{ id: ALL_PRODUCTS_ID, name: allProducts }];
-        quote.entries?.forEach((entry) => {
+        quote?.entries?.forEach((entry: OrderEntry) => {
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/details/comment/quote-details-comment.component.ts]:
```suggestion
        quote.entries?.forEach((entry: OrderEntry) => {
```
also the following chnages

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts [hunk 13]
```diff
@@ -154,13 +154,32 @@ describe('QuoteActionsByRoleComponent', () => {
...
+    });
+  });
+
   it('should read quote details state', (done) => {
     component.quoteDetails$.pipe(take(1)).subscribe((state) => {
       expect(state).toEqual(mockQuote);
       done();
     });
   });
-
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts]: i think now you not just removed the empty lines you added, but also additional ones. Can you please restore the formatting, so there are no empty line changes?  May just copy the file from the epic and re-add (copy from your change the test)? thx.

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts [hunk 14]
```diff
@@ -154,13 +154,32 @@ describe('QuoteActionsByRoleComponent', () => {
     expect(facade).toBeDefined();
   });
 
+  xdescribe('Ghost animation', () => {
```
Comment (Uli-Tiger) [inline feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts]:
```suggestion
  describe('Ghost animation', () => {
```

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts (hunk 13 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts]: Done!

File: feature-libs/quote/components/details/comment/quote-details-comment.component.ts (hunk 12 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.ts]: Done

File: feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts (hunk 11 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts]: Done!

Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts]: done

File: feature-libs/quote/components/quote-list/quote-list.component.spec.ts (hunk 9 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-list/quote-list.component.spec.ts:37]: Done

File: feature-libs/quote/components/quote-list/quote-list.component.spec.ts (hunk 7 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-list/quote-list.component.spec.ts:56]: done

File: feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts (hunk 6 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/quote-actions-by-role/quote-actions-by-role.component.spec.ts]: done

File: feature-libs/quote/components/details/cart/quote-details-cart.component.spec.ts (hunk 5 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/cart/quote-details-cart.component.spec.ts]: done

File: feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts (hunk 4 above)
Comment (Larisa-Staroverova) [reply feature-libs/quote/components/details/comment/quote-details-comment.component.spec.ts]: done

**Review by Uli-Tiger — APPROVED** [review]
